File: Kotlin/test/appendix/ThreadSafeTests.kt
package appendix

import appendix.ThreadSafe.SafeSet1
import appendix.ThreadSafe.SafeSet2
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import kotlin.test.Test
import kotlin.test.assertEquals

class ThreadSafeTests {
  @Test
  fun testSafeSet1() {
    testSafeSet(1_000_000, SafeSet1.SafeSet<Int>(), { set, x -> set.add(x) }, { it.all() })
  }

  @Test
  fun testSafeSet2() {
    testSafeSet(10_000, SafeSet2.SafeSet<Int>(), { set, x -> set.add(x) }, { it.all() })
  }

  private fun <A> testSafeSet(n: Int, set: A, add: (A, Int) -> Int, all: (A) -> Set<Int>) {
    val exec = Executors.newCachedThreadPool()
    val nums = (1..n).toSet()
    val start = CountDownLatch(2)

    val f1 = exec.submit {
      start.countDown()
      start.await()
      for (num in 1..n)
        assertEquals(num, add(set, num))
    }

    val f2 = exec.submit {
      start.countDown()
      start.await()
      while (true) {
        val s = all(set)
        assert(nums.containsAll(s))
        if (s.size == nums.size) break
      }
    }

    exec.shutdown()
    f1.get()
    f2.get()
  }
}

File: Scala/test/appendix/ThreadSafeTests.scala
package appendix

import org.scalatest.funsuite.AnyFunSuite
import tinyscalautils.assertions.in
import tinyscalautils.threads.Executors.global
import tinyscalautils.threads.{ countDownAndWait, withThreadPoolAndWait }

import java.util.concurrent.CountDownLatch
import scala.annotation.tailrec
import scala.concurrent.Future
import scala.jdk.CollectionConverters.SetHasAsScala

class ThreadSafeTests extends AnyFunSuite:
   test("SafeSet (Scala)") {
      val range = 1 to 1_000_000
      withThreadPoolAndWait(global) {
         val set   = ThreadSafe.SafeSet[Int]()
         val start = CountDownLatch(2)
         val f1 = Future {
            start.countDownAndWait()
            for i <- range do assertResult(i)(set += i)
         }
         val f2 = Future {
            start.countDownAndWait()
            @tailrec def loop(): Unit =
               val all = set.all
               assert(all.forall(_ in range))
               if all.size < range.size then loop()
            loop()
         }
         f1.zip(f2)
      }
   }

   test("SafeSet (Java)") {
      val range = 1 to 1_000_000
      withThreadPoolAndWait(global) {
         val set   = ThreadSafeJ.SafeSet[Int]()
         val start = CountDownLatch(2)
         val f1 = Future {
            start.countDownAndWait()
            for i <- range do assertResult(i)(set.add(i))
         }
         val f2 = Future {
            start.countDownAndWait()
            @tailrec def loop(): Unit =
               val all = set.all.asScala
               assert(all.forall(_ in range))
               if all.size < range.size then loop()
            loop()
         }
         f1.zip(f2)
      }
   }
